File: packages/core/src/commands/fs.ts
import type { Content } from '../content.ts'
import { displayName, resolveDir, resolveFile } from '../content.ts'
import type { Line, Output } from '../output.ts'
import { emptyOutput, failureOutput, plainLine, styledLine, textOutput } from '../output.ts'
import type { Session } from '../session.ts'
import { lines } from '../text.ts'

/**
 * Rust's `trim_start_matches('/')` removes ALL leading slashes, not just one.
 * Hence the `+` quantifier.
 */
export function resolvePath(s: Session, arg: string): string[] {
  const path = arg.startsWith('/') ? [] : [...s.cwd]
  for (const segment of arg.replace(/^\/+/, '').split('/')) {
    if (segment === '' || segment === '.') continue
    if (segment === '..') path.pop()
    else path.push(segment)
  }
  return path
}

export function pwd(s: Session): Output {
  return textOutput(`~/${s.cwd.join('/')}`)
}

export function ls(s: Session, c: Content, args: string[]): Output {
  const showAll = args.includes('-a')
  const named = args.find((a) => !a.startsWith('-'))
  const target = named === undefined ? [...s.cwd] : resolvePath(s, named)

  const dir = resolveDir(c, target)
  if (!dir) return failureOutput(`ls: ${target.join('/')}: no such directory`)

  const out: Line[] = dir.dirs.map((d) => styledLine(`${d.name}/`, 'accent'))

  const shown = dir.files.filter((f) => showAll || !f.hidden)

  // Titles align to the longest name in the directory. Three fixed spaces were
  // enough while the content was the personal portfolio, where the two names
  // happened to be nearly the same length. On any company journey it produced
  // staircase columns.
  const width = Math.max(0, ...shown.map((f) => displayName(f).length))

  for (const file of shown) {
    out.push({
      spans: [
        { text: displayName(file).padEnd(width), style: 'plain' },
        { text: `   ${file.title}`, style: 'dim' },
      ],
    })
  }

  if (out.length === 0) return textOutput('(empty)')
  return { lines: out, failed: false }
}

export function cd(s: Session, c: Content, args: string[]): Output {
  const arg = args[0]
  if (arg === undefined) {
    s.cwd = []
    return emptyOutput()
  }
  const target = resolvePath(s, arg)
  if (!resolveDir(c, target)) return failureOutput(`cd: ${arg}: no such directory`)
  s.cwd = target
  return emptyOutput()
}

export function cat(s: Session, c: Content, args: string[]): Output {
  const arg = args[0]
  if (arg === undefined) return failureOutput('cat: which file? try ls')

  const path = resolvePath(s, arg)
  if (resolveDir(c, path)) return failureOutput(`cat: ${arg}: is a directory`)

  const file = resolveFile(c, path)
  if (!file) return failureOutput(`cat: ${arg}: no such file`)

  if (file.hidden) s.eggFound = true
  if (!s.readFiles.includes(file.name)) s.readFiles.push(file.name)

  return { lines: lines(file.body).map(plainLine), failed: false }
}

File: crates/core/src/commands/fs.rs
use crate::content::Content;
use crate::output::{Line, Output, Style};
use crate::session::Session;

pub fn resolve(session: &Session, arg: &str) -> Vec<String> {
    let mut path = if arg.starts_with('/') {
        Vec::new()
    } else {
        session.cwd.clone()
    };
    for segment in arg.trim_start_matches('/').split('/') {
        match segment {
            "" | "." => {}
            ".." => {
                path.pop();
            }
            other => path.push(other.to_string()),
        }
    }
    path
}

pub fn pwd(session: &Session) -> Output {
    Output::text(&format!("~/{}", session.cwd.join("/")))
}

pub fn ls(session: &Session, content: &Content, args: &[String]) -> Output {
    let show_all = args.iter().any(|a| a == "-a");
    let target: Vec<String> = match args.iter().find(|a| !a.starts_with('-')) {
        Some(arg) => resolve(session, arg),
        None => session.cwd.clone(),
    };

    let Some(dir) = content.resolve_dir(&target) else {
        return Output::failure(&format!("ls: {}: no such directory", target.join("/")));
    };

    let mut lines: Vec<Line> = dir
        .dirs
        .iter()
        .map(|d| Line::styled(&format!("{}/", d.name), Style::Accent))
        .collect();

    for file in &dir.files {
        if file.hidden && !show_all {
            continue;
        }
        lines.push(Line {
            spans: vec![
                crate::output::Span {
                    text: file.display_name(),
                    style: Style::Plain,
                },
                crate::output::Span {
                    text: format!("   {}", file.title),
                    style: Style::Dim,
                },
            ],
        });
    }

    if lines.is_empty() {
        return Output::text("(empty)");
    }
    Output {
        lines,
        ..Output::default()
    }
}

pub fn cd(session: &mut Session, content: &Content, args: &[String]) -> Output {
    let Some(arg) = args.first() else {
        session.cwd.clear();
        return Output::empty();
    };
    let target = resolve(session, arg);
    if content.resolve_dir(&target).is_none() {
        return Output::failure(&format!("cd: {arg}: no such directory"));
    }
    session.cwd = target;
    Output::empty()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::content::Content;
    use crate::session::Session;

    fn flatten(out: &Output) -> String {
        out.lines
            .iter()
            .map(|l| l.spans.iter().map(|s| s.text.as_str()).collect::<String>())
            .collect::<Vec<_>>()
            .join("\n")
    }

    #[test]
    fn ls_lists_directories_then_files() {
        let content = Content::load();
        let session = Session::new();
        let rendered = flatten(&ls(&session, &content, &[]));
        assert!(rendered.contains("projects/"));
        assert!(rendered.contains("whoami"));
    }

    #[test]
    fn ls_hides_the_egg_by_default() {
        let content = Content::load();
        let session = Session::new();
        let rendered = flatten(&ls(&session, &content, &[]));
        assert!(!rendered.contains(".knock"));
    }

    #[test]
    fn ls_dash_a_reveals_the_egg() {
        let content = Content::load();
        let session = Session::new();
        let rendered = flatten(&ls(&session, &content, &["-a".to_string()]));
        assert!(rendered.contains(".knock"));
    }

    #[test]
    fn cd_moves_and_pwd_reports() {
        let content = Content::load();
        let mut session = Session::new();
        cd(&mut session, &content, &["projects".to_string()]);
        assert_eq!(session.cwd, vec!["projects".to_string()]);
        assert!(flatten(&pwd(&session)).contains("~/projects"));
    }

    #[test]
    fn cd_dot_dot_goes_up_and_stops_at_the_root() {
        let content = Content::load();
        let mut session = Session::new();
        cd(&mut session, &content, &["projects".to_string()]);
        cd(&mut session, &content, &["..".to_string()]);
        assert!(session.cwd.is_empty());
        cd(&mut session, &content, &["..".to_string()]);
        assert!(session.cwd.is_empty(), "the root has no parent");
    }

    #[test]
    fn cd_into_nothing_explains_itself() {
        let content = Content::load();
        let mut session = Session::new();
        let rendered = flatten(&cd(&mut session, &content, &["nowhere".to_string()]));
        assert!(rendered.contains("no such directory"), "got: {rendered}");
        assert!(session.cwd.is_empty());
    }
}
